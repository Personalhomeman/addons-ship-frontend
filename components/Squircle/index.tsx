import React from 'react';
import { Image, BaseProps } from '@bitrise/bitkit';

interface Props extends BaseProps {
  fatness?: number;
  objectFit?: string;
}

export default ({ objectFit = 'covver', fatness = 0.02, ...props }: Props) => {
  const s = Math.min(Math.max(fatness, 0), 0.5);

  const path = `M .5,0
    C ${s},0 0,${s} 0,.5
    0,${1 - s} ${s},1 .5,1
    ${1 - s},1 1,${1 - s} 1,.5
    1,${s} ${1 - s},0 .5,0
    Z`;
  const id = `squircle-${fatness}`;

  return (
    <div>
      <svg width="0" height="0" style={{ display: 'block' }}>
        <defs>
          <clipPath id={id} clipPathUnits="objectBoundingBox">
            <path d={path} />
          </clipPath>
        </defs>
      </svg>

      <Image {...props} style={{ clipPath: `url(#${id})`, objectFit }} />
    </div>
  );
};
