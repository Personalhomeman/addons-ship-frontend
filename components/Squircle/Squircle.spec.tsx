import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Squircle from '.';

describe('Squircle', () => {
  it('renders without issues', () => {
    const tree = shallowToJson(shallow(<Squircle src="img.url" />));
    expect(tree).toMatchSnapshot();
  });
});
