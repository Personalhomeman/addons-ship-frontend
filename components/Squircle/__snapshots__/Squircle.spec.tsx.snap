// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Squircle renders without issues 1`] = `
<div>
  <svg
    height="0"
    width="0"
  >
    <defs>
      <clipPath
        clipPathUnits="objectBoundingBox"
        id="squircle-0.05"
      >
        <path
          d="M .5,0
    C 0.05,0 0,0.05 0,.5
    0,0.95 0.05,1 .5,1
    0.95,1 1,0.95 1,.5
    1,0.05 0.95,0 .5,0
    Z"
        />
      </clipPath>
    </defs>
  </svg>
  <Image
    src="img.url"
    style={
      Object {
        "clipPath": "url(#squircle-0.05)",
        "marginTop": "-25%",
        "objectFit": "covver",
      }
    }
  />
</div>
`;
